### PR TITLE
Keep track of positions of each builder in sir_cx.

### PR DESCRIPTION
**Pull request by ptersilie**
> Requires https://github.com/softdevteam/yk/pull/44 to be merged first.

Whenever LLVM updates the position of a builder via
`LLVMPositionBuilderBefore` or `LLVMPositionBuilderAtEnd`, we need to
cache the basic block and index in sir_cx to allow us to insert
instructions at the correct position in the SIR blocks.

### CODE REVIEW

**Comment by vext01** [comment]
The PR and commit message has a typo:
> allow us insertion instructions at the correct position

File: src/librustc/sir.rs [hunk 1]
```diff
@@ -55,6 +58,8 @@ pub struct SirCx {
     pub llvm_blocks: FxHashMap<*const BasicBlock, (SirFuncIdx, SirBlockIdx)>,
     /// Function store. Also owns the blocks
     pub funcs: IndexVec<SirFuncIdx, ykpack::Body>,
+    /// Stores the current position (BasicBlock, Instruction index) for each builder.
```
Comment (vext01) [inline src/librustc/sir.rs]: I think we can shorten this to something like:
> Mirrors the insertion point for each LLVM `IrBuilder`

There's probably no need to re-state the types in the comment.

File: src/librustc/sir.rs [hunk 2]
```diff
@@ -137,4 +143,33 @@ impl SirCx {
 
         Ok(())
     }
+
+    /// Given an llvm::BasicBlock returns the equivalent ykpack::BasicBlock in the SIR.
+    fn get_sir_block(&mut self, bb: *const BasicBlock) -> &mut ykpack::BasicBlock {
+        let (sirfuncidx, sirblockidx) = self.llvm_blocks[&bb];
+        let sir_func = &mut self.funcs[sirfuncidx];
+        let sir_block = &mut sir_func.blocks[sirblockidx.index()];
```
Comment (vext01) [inline src/librustc/sir.rs]: May aswell get rid of the assignment, since we just return immediately.

Comment (vext01) [reply src/librustc/sir.rs]: Should we backtick all references to variables and types?

e.g.
```
/// Given an `llvm::BasicBlock` returns the equivalent `ykpack::BasicBlock` in the SIR.
```

File: src/librustc/sir.rs [hunk 3]
```diff
@@ -137,4 +143,33 @@ impl SirCx {
...
+        sir_block
+    }
+
+    /// Set the current position of builder to `pos`. Equivalent to LLVMPositionBuilderBefore.
```
Comment (vext01) [inline src/librustc/sir.rs:154]: I like that we are pointing out the LLVM equivalents :)

File: src/librustc_codegen_llvm/base.rs [hunk 4]
```diff
@@ -192,12 +193,17 @@ pub fn compile_codegen_unit(
...
                             .unwrap();
-                            let mut bx = Builder::with_cx(&cx);
-                            bx.position_at_end(llbb);
-                            bx.add_yk_block_label(llbb, lbl_name);
+                            labels.push((llbb, lbl_name));
                         }
                     }
                 }
+                // Only apply labels after collecting them in the previous step to avoid a double
+                // borrow of sir_cx
```
Comment (vext01) [inline src/librustc_codegen_llvm/base.rs]: Add a full-stop.

Comment (vext01) [reply src/librustc_codegen_llvm/base.rs]: And if we want to be totally precise, the problem is a double *mutable* borrow. Multiple immutable borrows are allowed.

File: src/rustllvm/RustWrapper.cpp [hunk 5]
```diff
@@ -1000,6 +1000,20 @@ extern "C" bool LLVMRustAddYkBlockLabel(LLVMBuilderRef Builder,
     return true;
 }
 
+extern "C" unsigned int LLVMRustInstructionIndex(LLVMValueRef Value) {
+    Instruction *Instr = unwrap<Instruction>(Value);
+    BasicBlock *B = Instr->getParent();
+    unsigned int index = 0;
+    for (BasicBlock::iterator I = B->begin(), IE = B->end(); I != IE; ++I) {
+        Instruction *CI = cast<Instruction>(I);
```
Comment (vext01) [inline src/rustllvm/RustWrapper.cpp:1008]: I think this can be a `static_cast`, which is more efficient, as it eliminates a type check.

https://stackoverflow.com/questions/28002/regular-cast-vs-static-cast-vs-dynamic-cast

File: src/rustllvm/RustWrapper.cpp [hunk 6]
```diff
@@ -1000,6 +1000,20 @@ extern "C" bool LLVMRustAddYkBlockLabel(LLVMBuilderRef Builder,
     return true;
 }
 
+extern "C" unsigned int LLVMRustInstructionIndex(LLVMValueRef Value) {
```
Comment (vext01) [inline src/rustllvm/RustWrapper.cpp:1003]: This method returns the wrong result if the instruction is not found.

We should either add a comment that states that the instruction must be present, or we can make this crash or return an error if it isn't found.

File: src/rustllvm/RustWrapper.cpp (hunk 5 above)
Comment (ptersilie) [reply src/rustllvm/RustWrapper.cpp:1008]: Okay, so I just replace `cast` with `static_cast`. Makes me wonder why the llvm code doesn't use `static_cast`s then.

File: src/rustllvm/RustWrapper.cpp (hunk 6 above)
Comment (ptersilie) [reply src/rustllvm/RustWrapper.cpp:1003]: I don't see how that can happen. I don't think there can be an instruction without a parent basic block. And if there is a parent basic block, then we will find that instruction inside it.

File: src/rustllvm/RustWrapper.cpp (hunk 5 above)
Comment (vext01) [reply src/rustllvm/RustWrapper.cpp:1008]: That's a good question.

File: src/rustllvm/RustWrapper.cpp (hunk 6 above)
Comment (vext01) [reply src/rustllvm/RustWrapper.cpp:1003]: You are right. My bad.

File: src/rustllvm/RustWrapper.cpp (hunk 5 above)
Comment (ptersilie) [reply src/rustllvm/RustWrapper.cpp:1008]: Let's leave it for now. We can always optimise later. I reckon this whole code is gonna change anyways at some point.

Comment (vext01) [reply src/rustllvm/RustWrapper.cpp:1008]: I agree.

All of the docs I've read don't even mention the cast operators used in this file.

I wonder if `cast` and `dyn_cast` are `static_cast` and `dynamic_cast` renamed...

Comment (ptersilie) [reply src/rustllvm/RustWrapper.cpp:1008]: That's certainly possible.

File: src/librustc_codegen_llvm/base.rs [hunk 7]
```diff
@@ -198,7 +198,7 @@ pub fn compile_codegen_unit(
                     }
                 }
                 // Only apply labels after collecting them in the previous step to avoid a double
-                // borrow of sir_cx
+                // borrow of sir_cx.
```
Comment (vext01) [inline src/librustc_codegen_llvm/base.rs]: double *mutable* borrow...

Comment (ptersilie) [reply src/librustc_codegen_llvm/base.rs]: Fixed.

**Comment by ptersilie** [comment]
Let me when I should squash.

**Comment by vext01** [comment]
Please squash

**Comment by ptersilie** [comment]
Squashed!

**Comment by vext01** [comment]
Please fix the typo in the commit message and the PR description. (Both appear in git history).

**Comment by ptersilie** [comment]
Fixed.

**Comment by vext01** [comment]
OK, so we need a three phase merge.

Please push a cycle breaker commit (https://softdevteam.github.io/ykdocs/misc/ci-cycles.html)

**Comment by ptersilie** [comment]
You sure? I think we can just merge https://github.com/softdevteam/yk/pull/45 first and it should work, since the changes in https://github.com/softdevteam/yk/pull/45 don't depend on the changes in here.

**Comment by vext01** [comment]
We can try, but the old compiler the CI has cached will probably choke on the revised ykpack types.

**Comment by ptersilie** [comment]
I don't think I follow. There's nothing in this PR that would change this.

**Comment by ptersilie** [comment]
I don't think I follow. There's nothing in this PR that would change this.

**Comment by ptersilie** [comment]
There you go, it worked.

**Comment by ptersilie** [comment]
There you go, it worked.

**Comment by vext01** [comment]
We got lucky it seems.

So:
bors r+

**Comment by ptersilie** [comment]
> We got lucky it seems.

We didn't introduce any new cyclic dependencies so "luck" is not the word I'd use.

**Comment by vext01** [comment]
bors r+

(got lost in github outage?)

**Comment by vext01** [comment]
bors ping

**Comment by vext01** [comment]
The new code-style checker got ya!

```
Diff in /home/buildbot-worker1/worker1/ykrustc-buildbot-build-script-hwt/build/src/librustc_codegen_llvm/builder.rs at line 154:
             llvm::LLVMPositionBuilderBefore(self.llbuilder, instr);
         }
 
-        let pos = unsafe {
-            llvm::LLVMRustInstructionIndex(instr)
-        };
+        let pos = unsafe { llvm::LLVMRustInstructionIndex(instr) };
 
-        let bb = unsafe {
-            llvm::LLVMGetInstructionParent(instr)
-        } as *const llvm::BasicBlock as *const sir::BasicBlock;
+        let bb = unsafe { llvm::LLVMGetInstructionParent(instr) } as *const llvm::BasicBlock
+            as *const sir::BasicBlock;
 
         let builder = self.llbuilder as *const llvm::Builder<'_> as *const sir::Builder;
```

Run `./x.py test` to see these errors (part of tidy).

**Comment by ptersilie** [comment]
This should fix it.

**Comment by vext01** [comment]
Please squash

**Comment by ptersilie** [comment]
Squashed!

**Comment by ptersilie** [comment]
Huh, that's weird. It claims the type we added in ykpack doesn't exist even though that PR was already merged.

**Comment by ptersilie** [comment]
bors try

**Comment by ptersilie** [comment]
Okay, let's give this a go before we squash:

bors try

**Comment by vext01** [comment]
Please squash.

**Comment by ptersilie** [comment]
Squashed!
